py/sentry_streams/flink/flink_adapter.py

```diff
@@ -81,5 +91,45 @@ def map(self, step: Map, stream: Any) -> Any:
         else:
             imported_fn = step.function
 
+        return_type = imported_fn.__annotations__["return"]
         # TODO: Ensure output type is configurable like the schema above
-        return stream.map(func=lambda msg: imported_fn(msg), output_type=Types.STRING())
+        return stream.map(
+            func=lambda msg: imported_fn(msg), output_type=FLINK_TYPE_MAP[return_type]
+        )
+
+    def reduce(self, step: Reduce, stream: Any) -> Any:
+
+        agg: Accumulator = step.aggregate_fn
+        windowing: Window = step.windowing
+        flink_window = FlinkWindows(windowing)
+        window_assigner = flink_window.build_window()
+        trigger = flink_window.get_trigger()
+
+        # The only optional parameter
+        group_by = step.group_by_key
+
+        # TODO: Configure WatermarkStrategy as part of KafkaSource
+        # Injecting strategy within a step like here produces
+        # a new, watermarked stream
+        watermark_strategy = WatermarkStrategy.for_monotonous_timestamps()
+        time_stream = stream.assign_timestamps_and_watermarks(watermark_strategy)
+
+        if group_by:
+            group_by_key = step.group_by_key
+            assert group_by_key is not None
+
+            keyed_stream = time_stream.key_by(FlinkGroupBy(group_by_key))
+
+            windowed_stream = keyed_stream.window(window_assigner).trigger(trigger)
+
+        else:
+            windowed_stream = time_stream.window_all(window_assigner).trigger(trigger)
+
+        return_type = agg.get_output.__annotations__["return"]
+
+        # TODO: Figure out a systematic way to convert types
+        return windowed_stream.aggregate(
+            FlinkAggregate(agg),
+            accumulator_type=Types.TUPLE([Types.STRING(), Types.INT()]),
```

Comment:
Could accumulators collect other aggregation types as well?

Reply:
For sure. We need to figure out a way to sytematically convert Python types into types that Flink understands. 

Hence the TODO comment